src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -475,12 +479,58 @@ ReturnRowsFromTuplestore(uint64 tupleCount, TupleDesc tupleDescriptor,
  * connection so that we receive results a row at a time.
  */
 static bool
-SendQueryInSingleRowMode(PGconn *connection, char *query)
+SendQueryInSingleRowMode(PGconn *connection, char *query, ParamListInfo paramListInfo)
 {
 	int querySent = 0;
 	int singleRowMode = 0;
 
-	querySent = PQsendQuery(connection, query);
+	if (paramListInfo != NULL)
+	{
+		int parameterIndex = 0;
+		int parameterCount = paramListInfo->numParams;
+		Oid *parameterTypes = palloc(parameterCount * sizeof(Oid));
+		const char **parameterValues = palloc(parameterCount * sizeof(char *));
+
+		/* get parameter types and values */
+		for (parameterIndex = 0; parameterIndex < parameterCount; parameterIndex++)
+		{
+			ParamExternData *parameterData = &paramListInfo->params[parameterIndex];
+			Oid typeOutputFunctionId = InvalidOid;
+			bool variableLengthType = false;
+
+			/*
+			 * Use text oid for data types where the oid values can be different
+			 * on the master and worker nodes.
+			 */
+			if (parameterData->ptype >= FirstNormalObjectId)
+			{
+				parameterTypes[parameterIndex] = TEXTOID;
```

Comment:
> Hm. I think this should be 0 instead, which means the type should be automatically inferred. There's not necessarily a cast from text to the other side of the operation, so using text isn't a good idea here.

Using 0 was the my initial attempt looking into the [documentation](https://www.postgresql.org/docs/9.5/static/libpq-exec.html) of PQexecParams(). However, if I use 0, this errors out at the even first trial on the worker node.

```
ERROR:  could not determine data type of parameter $1
STATEMENT:  INSERT INTO helloworld_102225 (id, set) VALUES (1, '\x118b7f'::hll)
```

There are a couple of places where this error message is used in PostgreSQL source. I will check which one is fired, and try to find a solution.